### PR TITLE
Changes to the FileDiff section of Differ to support HLV comparison

### PR DESCRIPTION
**Pull request by darshan-na**
The changes include the following
1) Separating the Xattr section from the body 
2) Writing the Xattrs and the body independently to the file
3) Reading the Xattrs and the body seperatly
4) Creating an Iterator over the Xattrs to fetch the HLV
5) Creating the HLV struct and storing it as a part of the oneEntry struct that contains all the necessary fields for comparison of docs
6) adding the compareHlv function that does the HlvComparison

### CODE REVIEW

File: dcp/DcpHandler.go [hunk 1]
```diff
@@ -525,9 +525,24 @@ func (m *Mutation) ToUprEvent() *xdcrBase.WrappedUprEvent {
...
 //	(per col filter) - 2 byte
 func (mut *Mutation) Serialize() []byte {
+	var bodyHash [64]byte
+	var xattrSize uint32
+	var xattr []byte
+
+	if mut.Datatype&xdcrBase.XattrDataType > 0 {
+		bodyWithoutXattr, err := xdcrBase.StripXattrAndGetBody(mut.Value)
+		if err != nil {
+			fmt.Println("Error encountered while stripping Xattrs from the body, err: ", err)
```
Comment (nelio2k) [inline dcp/DcpHandler.go]: Log the doc key

File: main.go [hunk 2]
```diff
@@ -763,8 +765,8 @@ func (difftool *xdcrDiffTool) diffDataFiles() error {
 		return fmt.Errorf("Error mkdir fileDifferDir: %v\n", err)
 	}
 
-	difftoolDriver := differ.NewDifferDriver(options.sourceFileDir, options.targetFileDir, options.fileDifferDir,
-		base.DiffKeysFileName, int(options.numberOfWorkersForFileDiffer), int(options.numberOfBins),
+	difftoolDriver := differ.NewDifferDriver(difftool.bucketTopologySvc, difftool.specifiedSpec, options.sourceFileDir, options.targetFileDir, options.fileDifferDir,
```
Comment (nelio2k) [inline main.go]: Add new arguments at the end, not at the beginning

File: differ/differDriver.go [hunk 3]
```diff
@@ -21,12 +21,44 @@ import (
...
 
+type pruningWindow struct {
+	duration time.Duration
+	lock     sync.RWMutex
+}
+
+var sourcePruningWindow *pruningWindow = &pruningWindow{}
+
+func (p *pruningWindow) set(svc service_def.BucketTopologySvc, spec *metadata.ReplicationSpecification) error {
+	notificationCh, err := svc.SubscribeToLocalBucketFeed(spec, "")
```
Comment (nelio2k) [inline differ/differDriver.go]: Should pass in a generated subscriber ID for completeness

File: differ/differDriver.go [hunk 4]
```diff
@@ -21,12 +21,44 @@ import (
...
+		fmt.Printf("Failed to fetch LocalBucketFeed. err=%v\n", err)
+		return err
+	}
+	defer svc.UnSubscribeLocalBucketFeed(spec, "")
+	latestNotification := <-notificationCh
+	defer latestNotification.Recycle()
+	pruningWindow := latestNotification.GetVersionPruningWindowHrs()
+	p.lock.Lock()
+	defer p.lock.Unlock()
+	p.duration = time.Duration(uint32(pruningWindow)*60*60) * time.Second
```
Comment (nelio2k) [inline differ/differDriver.go]: Simply do
`time.Duration(uint32(pruningWindow)) * time.Hour`

File: differ/differ.go [hunk 5]
```diff
@@ -232,6 +345,62 @@ func NewFilesDifferWithFDPool(file1, file2 string, fdPool *fdp.FdPool, collectio
...
+	var xattrHlv, xattrImportCas []byte
+	for XattrIterator.HasNext() {
+		key, value, err = XattrIterator.Next()
+		if err != nil {
+			return
+		}
+		if xdcrBase.Equals(key, xdcrBase.XATTR_HLV) {
+			xattrHlv = value
+		}
+		if xdcrBase.Equals(key, xdcrBase.XATTR_IMPORTCAS) {
```
Comment (nelio2k) [inline differ/differ.go]: This will change per @sumukhbhat2701 's changes on `_mou`

File: differ/differ.go [hunk 6]
```diff
@@ -232,6 +345,62 @@ func NewFilesDifferWithFDPool(file1, file2 string, fdPool *fdp.FdPool, collectio
...
+			xattrImportCas = value
+		}
+	}
+	// any errors in parsing importCas we dont parse/construct the hlv beacuse it is not possible to determine if the hlv is outdated/updated
```
Comment (nelio2k) [inline differ/differ.go]: You can still have HLV even if `importCAS` is not there, no?
HLV is out of date if the `cvCAS != CAS`

@sumukhbhat2701  is working on a mou change where the importCAS can be removed but HLV still exists

File: differ/differ.go [hunk 7]
```diff
@@ -232,6 +345,62 @@ func NewFilesDifferWithFDPool(file1, file2 string, fdPool *fdp.FdPool, collectio
...
+	} else {
+		Hlv, err2 = hlv.NewHLV(bucketId, docCas, cvCas, cvSrc, cvVer, pvMap, mvMap)
+	}
+	if err2 != nil {
+		return nil, err2
+	}
+	return Hlv, nil
+}
+
+func extractHlvAndImportCas(XattrIterator *xdcrBase.XattrIterator, bucketUUID string, docCas uint64) (hlv *hlv.HLV, importCas uint64, err error) {
```
Comment (nelio2k) [inline differ/differ.go]: Is there a reason why you tie HLV and ImportCAS together in this function?

File: differ/differ.go [hunk 8]
```diff
@@ -155,21 +167,122 @@ func (entry oneEntry) Diff(other oneEntry) (int, bool) {
...
+		return false
+	} else if item1.GetCvSrc() != item2.GetCvSrc() || item1.GetCvVer() != item2.GetCvVer() {
+		return false
+	} else if !comparePv(item1.GetPV(), item2.GetPV(), item1Cas, item2Cas) {
+		return false
+	}
+	return true
+}
+
+func comparePv(Pv1 hlv.VersionsMap, Pv2 hlv.VersionsMap, cas1 uint64, cas2 uint64) bool {
```
Comment (nelio2k) [inline differ/differ.go:223]: This function seems to be a prime candidate to write unit test cases for.
Try writing some unit tests and test boundary cases for this in `differ_test.go` as part of this commit

File: differ/differ.go [hunk 9]
```diff
@@ -155,21 +167,122 @@ func (entry oneEntry) Diff(other oneEntry) (int, bool) {
...
+		if hlv2.GetCvSrc() != bucketUUID1 || hlv2.GetCvVer() != cas1 {
+			return false
+		} else {
+			hlv1, _ = hlv.NewHLV(bucketUUID1, cas1, 0, "", 0, nil, nil)
+		}
+	}
+	return actualCompare(hlv1, hlv2, cas1, cas2)
+}
+
+func actualCompare(item1 *hlv.HLV, item2 *hlv.HLV, item1Cas uint64, item2Cas uint64) bool {
```
Comment (nelio2k) [inline differ/differ.go]: Better naming

File: differ/differ.go [hunk 10]
```diff
@@ -155,21 +167,122 @@ func (entry oneEntry) Diff(other oneEntry) (int, bool) {
...
 		} else if entry.Datatype != other.Datatype {
 			return 0, false
+		} else if !compareHlv(entry.Hlv, other.Hlv, entry.Cas, other.Cas, entry.ImportCas, other.ImportCas, entry.bucketUUID, other.bucketUUID) {
+			return 0, false
 		}
 	}
 	return 0, true
 }
 
+func compareHlv(hlv1, hlv2 *hlv.HLV, cas1, cas2, importCas1, importCas2 uint64, bucketUUID1, bucketUUID2 hlv.DocumentSourceId) bool {
```
Comment (nelio2k) [inline differ/differ.go]: Just looking at the code, I can see that it could be problematic if HLV ever has changes.

This function seems to me would be better implemented in the XDCR "library"
For the `type CRMetadata struct` in `crMetadata.go`, add a new `Diff(other CRMetadata)` method?

This function takes in:
- HLVs
- CAS's
- imporCAS's

Unsure about "bucketUUIDs"

But can you think about whether or not it makes sense to do it here or in the library?

File: dcp/DcpHandler.go [hunk 11]
```diff
@@ -525,9 +525,24 @@ func (m *Mutation) ToUprEvent() *xdcrBase.WrappedUprEvent {
 //	colFiltersLen - 2 byte (number of collection migration filters)
 //	(per col filter) - 2 byte
 func (mut *Mutation) Serialize() []byte {
+	var bodyHash [64]byte
+	var xattrSize uint32
+	var xattr []byte
+
+	if mut.Datatype&xdcrBase.XattrDataType > 0 {
```
Comment (nelio2k) [inline dcp/DcpHandler.go:546]: This change will strip away all Xattr. Is this what you intended? Or are you only supposed to take out HLV related xattr?

File: differ/differDriver.go (hunk 4 above)
Comment (darshan-na) [reply differ/differDriver.go]: But shouldn't it be in seconds for the comparison `xdcrBase.CasDuration(value1, cas) >= pruningWindow` ?

File: differ/differ.go (hunk 5 above)
Comment (darshan-na) [reply differ/differ.go]: Noted

File: differ/differ.go (hunk 6 above)
Comment (darshan-na) [reply differ/differ.go]: Yes, the existing logic takes care of reading hlv even if importCas is not present 
But if import is present and we encounter a parsing error, we shouldn't proceed to parse the HLV because it is not possible to determine if the HLV is outdated or not right? 
So I capture the error and prevent parsing the HLV

File: dcp/DcpHandler.go (hunk 11 above)
Comment (darshan-na) [reply dcp/DcpHandler.go:546]: I did think about it and eventually I decided to strip the Xattrs as whole and write it to file.

Reason : As of now the scope of the project is limited hlv comparison but in the future if at all we decide to compare other system Xattrs (say eventing,fts) it becomes clumsy to write/read each system Xattr independently to file.

My rationale was that by keeping the Xattrs as a whole separate, we can easily create an iterator to select the specific system Xattrs whenever needed.

However, this approach introduces a tradeoff: for body comparison, we would need to run an iterator to strip the HLVs and any other Xattrs that the user chooses to exclude from the comparison, taken as input from the command line.

File: differ/differ.go (hunk 8 above)
Comment (darshan-na) [reply differ/differ.go:223]: Sure, I'll write some unit tests

File: differ/differ.go (hunk 10 above)
Comment (darshan-na) [reply differ/differ.go]: Yeah,I'll think about it

File: differ/differ.go (hunk 7 above)
Comment (darshan-na) [reply differ/differ.go]: No, but I remember seeing it done that way on the XDCR side, so I preferred to follow it.

File: differ/differ.go (hunk 10 above)
Comment (nelio2k) [reply differ/differ.go]: @darshan-na  - just for clarity. What is your decision?

File: runDiffer.sh [hunk 12]
```diff
@@ -106,6 +106,9 @@ while getopts ":h:p:u:r:s:t:n:q:v:cbm:ew:d" opt; do
 	w)
 		setupTimeout=$OPTARG
 		;;
+	x)
```
Comment (nelio2k) [inline runDiffer.sh:105]: Update `printHelp()`

File: main.go [hunk 13]
```diff
@@ -620,7 +624,12 @@ func main() {
 		}
 	}
 	if options.runDataGeneration {
-		err := difftool.generateDataFiles()
+		xattrsNocompare := strings.Split(options.xattrsNoCompare, ":")
+		var xattrKeysForNoCompare map[string]bool = make(map[string]bool)
+		for _, key := range xattrsNocompare {
+			xattrKeysForNoCompare[key] = true
+		}
+		err := difftool.generateDataFiles(xattrKeysForNoCompare)
```
Comment (nelio2k) [inline main.go]: It would be preferred to not change the `generateDataFiles` call argument.
Is there a cleaner way to pass along this as part of the `difftool` constructor? 
Perhaps line 627-631 can be refactored into a global function and the function "pointer" can be passed along to the `difftool` constructor

File: main.go [hunk 14]
```diff
@@ -238,7 +241,8 @@ func argParse() {
 		"The differ to be run with debug log level and the SDK/gocb logging will also be enabled.")
 	flag.IntVar(&options.setupTimeout, "setupTimeout", base.SetupTimeoutSeconds,
 		"Common setup timeout duration in seconds")
-
+	flag.StringVar(&options.xattrsNoCompare, "xattrsNoCompare", "",
+		"A string denoting the Xattrs that shouldn't be considered for comparison, The induvidual Xattrs are seperated by \":\"")
```
Comment (nelio2k) [inline main.go]: 1. Typo: "individual"
2. Xattr may contain ":" as a valid character in its key. It may not be good to use it as a separator. See: https://couchbase.slack.com/archives/C968F6RDZ/p1479979092313248

It may be better to consider a few alternatives:
1. A list of pre-set xattrs that differ is able to understand, and then the -x can specify the union of these xattrs
i.e. 1. _mou 2. _eventing
-x 3 means 0x11 meaning both _mou and _eventing is specified
-x 2 means only _eventing is specified.

2. -x points to a flat file of xattrs keys to skip. Read the file and parse it.

File: differ/differDriver.go (hunk 4 above)
Comment (nelio2k) [reply differ/differDriver.go]: CAS isn't seconds.
See: https://go.dev/play/p/IsQXFtmpd_K

File: dcp/DcpClient.go [hunk 15]
```diff
@@ -296,7 +296,7 @@ func (c *DcpClient) initializeDcpHandlers() error {
 		dcpHandler, err := NewDcpHandler(c, c.dcpDriver.fileDir, i, vbList, c.dcpDriver.numberOfBins,
 			c.dcpDriver.dcpHandlerChanSize, c.dcpDriver.fdPool, c.dcpDriver.IncrementDocReceived,
 			c.dcpDriver.IncrementSysOrUnsubbedEventReceived, c.colMigrationFilters, c.utils, c.bufferCap,
-			c.migrationMapping)
+			c.migrationMapping, &xdcrBase.XattrIterator{})
```
Comment (nelio2k) [inline dcp/DcpClient.go]: Why do you need to pass in an empty xattr iterator?
Why not just create one inside the dcp handler?

File: dcp/DcpHandler.go [hunk 16]
```diff
@@ -221,7 +223,12 @@ func (dh *DcpHandler) processMutation(mut *Mutation) {
 	if dh.colMigrationFiltersOn && len(filterIdsMatched) > 0 {
 		mut.ColFiltersMatched = filterIdsMatched
 	}
-	bucket.write(mut.Serialize())
+	ret, err := mut.Serialize(dh.xattraIterator, dh.dcpClient.dcpDriver.xattrKeysForNoCompare)
```
Comment (nelio2k) [inline dcp/DcpHandler.go]: Simlarly, is there a way to not need to pass in arguments as part of `Serialize()`? `mut` can contain these from constructor, no?

File: dcp/DcpHandler.go [hunk 17]
```diff
@@ -577,5 +595,46 @@ func (mut *Mutation) Serialize() []byte {
 		binary.BigEndian.PutUint16(ret[pos:pos+2], uint16(colFilterId))
 		pos += 2
 	}
-	return ret
+	return ret, nil
+}
+
+func stripHlvAndImportFromXattr(xattr []byte, size int, xattrSize uint32, xattrIterator *xdcrBase.XattrIterator, keysToExclude map[string]bool, bodyWithoutXattr []byte) ([]byte, []byte, []byte, error) {
```
Comment (nelio2k) [inline dcp/DcpHandler.go]: Add comment on input values and output values, hard to read otherwise

This function should be more generic, something like `removeKVSubsetFromXattr`, and the `keysToExclude` should include either `XATTR_HLV` or `XATTR_IMPORTCAS`.
Then the function itself should have the return signature similar to:
1. trimmedXattr
2. A map of [key]value pair of the excluded xattr keys
3. error

File: dcp/DcpHandler.go [hunk 18]
```diff
@@ -524,25 +531,32 @@ func (m *Mutation) ToUprEvent() *xdcrBase.WrappedUprEvent {
...
+			return nil, err
 		}
-		bodyHash = sha512.Sum512(bodyWithoutXattr)
 		xattrSize, _ = xdcrBase.GetXattrSize(mut.Value)
 		xattr = mut.Value[4 : xattrSize+4]
+		trimmedXattrPlusBody, hlv, importCas, err = stripHlvAndImportFromXattr(xattr, len(mut.Value), xattrSize, xattraIterator, XattrsToExclude, bodyWithoutXattr)
+		if err != nil {
+			return nil, err
+		}
+		bodyHash = sha512.Sum512(trimmedXattrPlusBody)
```
Comment (nelio2k) [inline dcp/DcpHandler.go:560]: Can you double check that the xattrs are in sorted order?  You can check against KV team.

Is it possible that doc A has xattr in the order of <key1, value1> and <key2, value2>
Doc B has xattr <key2, value2>, <key1, value1>?

If the above is possible, is it possible then the SHA can be mismatched even though the document xattr is the same but just in different order?

File: differ/differ.go (hunk 10 above)
Comment (darshan-na) [reply differ/differ.go]: I think we can move it to the library. That way it is neater.
So I was thinking if we can add another field in the crMetadata struct for bucketUUID.

File: main.go (hunk 13 above)
Comment (darshan-na) [reply main.go]: Agreed, this version of the code was to simply test the logic. I have started working on refactoring the code to make such changes. Will push the changes soon.

File: main.go (hunk 14 above)
Comment (darshan-na) [reply main.go]: I think i prefer the second method over the first, because in future if there is another system Xattr that needs to be excluded that differ isn't aware of, then the second method comes handy.

File: dcp/DcpHandler.go (hunk 18 above)
Comment (darshan-na) [reply dcp/DcpHandler.go:560]: The KV team replied with, "Short answer is that you cannot depend on the ordering (if you try to fetch the SAME document multiple times the order would be the same, but if you do any modification to the document the ordering may or may not change)."
So i think we can stick to the current implementation of ordering the keys and then constructing the Xattrs based on the ordered keys.

File: differ/differDriver.go (hunk 4 above)
Comment (darshan-na) [reply differ/differDriver.go]: Yeah, understood

File: base/constants.go [hunk 19]
```diff
@@ -77,8 +77,8 @@ const KeyLenVariable = 2
 const MigrationFilterLen = 2
 const xattrSizeLen = 4
 
-func GetFixedSizeMutationLen(keyLen int, xattrSize uint32, colMigrationFilterMatched []uint8) int {
-	return KeyLenVariable + keyLen + xattrSizeLen + int(xattrSize) + BodyLength + MigrationFilterLen + len(colMigrationFilterMatched)*2
+func GetFixedSizeMutationLen(keyLen int, size uint32, colMigrationFilterMatched []uint8) int {
+	return KeyLenVariable + keyLen + (2 * xattrSizeLen) + int(size) + BodyLength + MigrationFilterLen + len(colMigrationFilterMatched)*2
```
Comment (nelio2k) [inline base/constants.go]: Why double the xattr size length? Can you add comment to explain

**Comment by nelio2k** [comment]
@darshan-na - you don't want to just have a "resolve comment" commit ([87f7e14](https://github.com/couchbaselabs/xdcrDiffer/pull/100/commits/87f7e14e40b822259fdfb54253331b022fe8522f))  that doesn't address the code until a later commit. It makes the reviewer's job harder.

**Comment by nelio2k** [comment]
It's hard to tell what the commit with the commit message "resolving comment" means.

It'd be good for you to either squash that commit or potentially break it up into smaller commits with commit messages that correspond to what you're actually fixing

File: dcp/DcpHandler.go [hunk 20]
```diff
@@ -260,15 +260,15 @@ func (dh *DcpHandler) SnapshotMarker(snapshot gocbcore.DcpSnapshotMarker) {
 }
 
 func (dh *DcpHandler) Mutation(mutation gocbcore.DcpMutation) {
-	dh.writeToDataChan(CreateMutation(mutation.VbID, mutation.Key, mutation.SeqNo, mutation.RevNo, mutation.Cas, mutation.Flags, mutation.Expiry, gomemcached.UPR_MUTATION, mutation.Value, mutation.Datatype, mutation.CollectionID))
+	dh.writeToDataChan(CreateMutation(mutation.VbID, mutation.Key, mutation.SeqNo, mutation.RevNo, mutation.Cas, mutation.Flags, mutation.Expiry, gomemcached.UPR_MUTATION, mutation.Value, mutation.Datatype, mutation.CollectionID, dh.xattraIterator, dh.dcpClient.dcpDriver.xattrKeysForNoCompare))
```
Comment (nelio2k) [inline dcp/DcpHandler.go]: `writeToDataChan` is a method of `dh`. `dh` itself has access to `dh.xattriterator` and `dh.dcpClient`. No need to re-add it to function signature

Also, xattrIterator is mis-spelled

File: dcp/DcpHandler.go [hunk 21]
```diff
@@ -290,17 +290,17 @@ func (dh *DcpHandler) FlushCollection(flush gocbcore.DcpCollectionFlush) {
 
 func (dh *DcpHandler) CreateScope(creation gocbcore.DcpScopeCreation) {
 	// Overloading collectionID field for scopeID because differ doesn't care
-	dh.writeToDataChan(CreateMutation(creation.VbID, nil, creation.SeqNo, 0, 0, 0, 0, gomemcached.DCP_SYSTEM_EVENT, nil, 0, creation.ScopeID))
+	dh.writeToDataChan(CreateMutation(creation.VbID, nil, creation.SeqNo, 0, 0, 0, 0, gomemcached.DCP_SYSTEM_EVENT, nil, 0, creation.ScopeID, nil, nil))
```
Comment (nelio2k) [inline dcp/DcpHandler.go:293]: These function signatures don't need to change per my comment above

File: dcp/DcpHandler.go (hunk 20 above)
Comment (darshan-na) [reply dcp/DcpHandler.go]: "Simlarly, is there a way to not need to pass in arguments as part of Serialize()? mut can contain these from constructor, no?"
From this comment, I have added the iterator and the map containing keys for noCompare as a part of the mut constructor.
In the above line of code i am add them to the constructor of the "mutation struct".
writeToDataChan function signature remains the same

File: base/constants.go (hunk 19 above)
Comment (darshan-na) [reply base/constants.go]: Sure

File: dcp/DcpHandler.go (hunk 21 above)
Comment (darshan-na) [reply dcp/DcpHandler.go:293]: As discussed, we don't change the function signature of `dh.writeToDataChan` . We are just passing the additional required parameters to the constructor of the Mutation struct.

**Review by nelio2k — APPROVED** [review]
